### ISSUE TITLE
Accessibility: Screen Reader Support

### ISSUE THREAD

**Issue by MJacred** [fyne-io/fyne#1093]
### Is your feature request related to a problem? Please describe:
To allow people who are blind, visually impaired, illiterate, etc. to use software.

### Is it possible to construct a solution with the existing API?
the desired outcome is not possible at this time

### Describe the solution you'd like to see:
* Linux: ATK/[AT-SPI](https://en.wikipedia.org/wiki/Assistive_Technology_Service_Provider_Interface)
* Windows: MSAA (Microsoft Active Accessibility) + [IA2 (IAccessible2)](https://en.wikipedia.org/wiki/IAccessible2)
* Apple: Universal Access (UA)

Like [Gecko](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Accessibility/AT-APIs)

**Comment by andydotxyz** [comment]
This is a great suggestion and something we need to add. Also for Android and iOS I’d course.
Any help appreciated :)

**Comment by ivan-avalos** [comment]
I wouldn't use Fyne or recommend it to anyone unless this gets fully implemented. It really should be top priority for every GUI framework.

**Comment by andydotxyz** [comment]
> I wouldn't use Fyne or recommend it to anyone unless this gets fully implemented. It really should be top priority for every GUI framework.

Thank you for your comment @ivan-avalos. If you are able to help this get implemented quicker through contributions or sponsorship we would be very grateful. Otherwise it will be completed as soon as we are able when balanced with other priorities.

**Comment by your-friend-alice** [comment]
this is also a critical feature for me to build anything serious with this library. just checking if this is still a priority in 2022. I can certainly try to contribute but that's pretty far outside my skillset.

**Comment by jspraul** [comment]
https://github.com/AccessKit/accesskit is a new hotness as accessibility infrastructure

**Comment by Bluebugs** [comment]
AccessKit is definitively on our radar and most likely the way forward. It does introduce additional native dependencies that will have to be solved. There is a lot of work here to be done and I think anyone help around this will be useful.

**Comment by DataTriny** [comment]
Hello @Bluebugs , AccessKit contributor here.

I know nothing about Go but since I came across this project I took a look at this issue. FYI we now have complete C bindings that can be called by Go through FFI.

If changing the build process is not an option, then checking out pre-built binaries into the repo might be the only solution. You can find the latest version [here](https://github.com/AccessKit/accesskit/releases/download/accesskit_c-v0.6.1/accesskit_c-v0.6.1.zip).

Otherwise, the build process for these bindings use CMake and require Rust and cbindgen to be available on the system.

**Comment by Bluebugs** [comment]
Hi @DataTriny , thanks for stopping by. Good to know there is C API now, this will make writing a Go wrapper easier. I think it will be best for the Go ecosystem if that wrapper was independent from this repository. fyne-io github organization could still host it, like we do for glfw-js for example, but maybe having it as a repository next to the AccessKit one make sense.

As for the integration with fyne itself, I think it should be an optional dependencies. We would add it to fyne-cross base image (our tool to build fyne for all platform in an easier way that use docker container to have a known build environment) and enable it by default when building via fyne-cross. That would be the least amount of burden and the easiest for everyone.

**Comment by DataTriny** [comment]
I agree with you that there is probably a lot of non fyne related code to write, an AccessKit Go bindings project would be desirable. But since there are no Go users of our bindings that I am aware of at this time, I am using fyne as a testbed.

> We would add it to fyne-cross base image and enable it by default when building via fyne-cross.

Since AccessKit is written in Rust, and since Go can't invoke Cargo to compile it, do you have an idea of how it could end up in a "fyne build"? When I do `go get fyne.io/fyne/v2@latest` it seems that Go just downloads the git repo, so I'm not sure what kind of build are you talking about? My understanding so far is that Go just builds all the dependencies when I do `go run` and then statically links everything in the final executable. Therefore I only see two options:

- Ask users to install AccessKit dynamic libraries on their system so that fyne apps can load them at runtime,
- Put pre-built AccessKit static libraries on a repo where `go get` can find them and provide appropriate compilation flags to cgo.

Am I missing something from the fyne development workflow? BTW I am only talking in the context of Windows, Linux and macOS since these are the only platforms we support at the moment.

**Comment by jspraul** [comment]
As a bystander, my best attempt at translation:

> add it to fyne-cross base image
adding rust dev tools to https://github.com/fyne-io/fyne-cross-images/blob/main/base/Dockerfile etc.
> building via fyne-cross
adding the rust build command to https://github.com/fyne-io/fyne-cross/blob/master/internal/command/linux.go#L164 etc.

So the recommended fyne-cross simplifying building native dependencies is used instead of `go get`.

The fun part is definitely supporting all the [operating systems x architectures](https://github.com/fyne-io/fyne-cross#fyne-cross).

**Comment by a-alhusaini** [comment]
Thought I'd stop by and mention that I'd be happy to test any a11y features fyne adds. I'm blind and would love to see more GUI libs add more a11y features!

**Comment by andydotxyz** [comment]
Thanks for that. We hope to get working on in it in the near future.

**Comment by dweymouth** [comment]
> I agree with you that there is probably a lot of non fyne related code to write, an AccessKit Go bindings project would be desirable. But since there are no Go users of our bindings that I am aware of at this time, I am using fyne as a testbed.
> 
> > We would add it to fyne-cross base image and enable it by default when building via fyne-cross.
> 
> Since AccessKit is written in Rust, and since Go can't invoke Cargo to compile it, do you have an idea of how it could end up in a "fyne build"? When I do `go get fyne.io/fyne/v2@latest` it seems that Go just downloads the git repo, so I'm not sure what kind of build are you talking about? My understanding so far is that Go just builds all the dependencies when I do `go run` and then statically links everything in the final executable. Therefore I only see two options:
> 
> * Ask users to install AccessKit dynamic libraries on their system so that fyne apps can load them at runtime,
> * Put pre-built AccessKit static libraries on a repo where `go get` can find them and provide appropriate compilation flags to cgo.
> 
> Am I missing something from the fyne development workflow? BTW I am only talking in the context of Windows, Linux and macOS since these are the only platforms we support at the moment.

I think # 2 would be the best option, with the caveat that we'd need static libs for every supported combination of OS + architecture. And as @Bluebugs mentioned it should probably be an optional dependency

**Comment by andydotxyz** [comment]
I'm not sure about optional here - that would lead to apps that could be accessible having that turned off by default.
Feels like a bad experience.

**Comment by dweymouth** [comment]
I guess an alternative could be embedding the static compiled libs for AccessKit into the fyne tool? Ensuring that they're always available to link against would be the challenge of making it a required dependency. Either that or rewrite AccessKit in Go or C I guess, since we can't introduce the rust compiler as a build dependency.

**Comment by andydotxyz** [comment]
Yes, those approaches are better.

**Comment by andydotxyz** [comment]
We have now merged accessibility in terms of screen reader for all OS - you can test and help us improve by using "-tags accessibility". This will be in v2.8.0, but disabled by default

**Comment by DataTriny** [comment]
I just tried the demo app in this repo and could only see the root window in the accessibility tree on Windows. I can already tell that the GLFW window is shown before the window procedure is swapped, which is known to cause issues, but is unlikely to be the main problem.

Other projects (Godot (C++), DVUI (Zip)...) were able to integrate AccessKit by relying on the static libraries we provide, avoiding any Rust compilation so this could have been an option here as well. I hope the team have carefully weighed this inconvenience against the huge amount of work required by a proper accessibility tree implementation.

In any case, I wish the Fyne project success in this endeavour. Congratulations for what have been achieved so far.

**Comment by andydotxyz** [comment]
I didn't manage to get AccessKit integrated, so took another path. We would happily accept contributions that resolve the shortcomings of the work in progress if you are interested.

As long as the licenses are compatible and the build process is no more complex I imagine it would be beneficial.
